File: data_fits/gridsamp.jl
function consecutive(f, A::Matrix{Float64})::Matrix{Float64}
    dims = size(A)
    df = Matrix{Float64}(undef,dims[1]-1,dims[2])
    for j = 1:dims[2]
        df[:,j] = [ f(A[i+1,j], A[i,j]) for i = 1:dims[1]-1 ]
    end
    return df
end

function gridsamp(bounds::Matrix{Float64}, q::Vector{Int64})
    mr, n = size(bounds)
    dr = consecutive(-,bounds)
    
    if mr != 2 || any(x->x<0, dr)
        throw(DomainError(bounds,"bounds must be an array with two rows and bounds(1,:) <= bounds(2,:)"))
    end
    
    if ndims(q) > 1 || any(x->x<=0, q)
        throw(DomainError(q,"q must be a vector with non-negative elements"))
    end
    
    p = length(q)
    if p == 1
        q = repeat(q, outer = (1, n))
    elseif p != n
        throw(DomainError(q,"length of q must be either 1 or $n"))
    end
    
    # Check for degenerate intervals
    i = findall(x->x==0, dr)
    if length(i) > 0
        for index in i
            q[index[2]] = 0 * q[index[2]]
        end
    end

    # Recursive computation
    if n > 1
        a = gridsamp(bounds[:, 2:end], q[2:end])  # Recursive call
        m,_ = size(a)
        q = q[1]

        s = hcat(zeros(m * q, 1), repeat(a, outer = (q, 1)))
        y = LinRange(bounds[1, 1], bounds[2, 1], q)

        k = 1:m
        for i = 1:q
            aug = fill(y[i], (m, 1))
            aug = reshape(aug, size(s[k, 1]))

            s[k, 1] = aug
            k = k .+ m
        end
    else
        s = LinRange(bounds[1, 1], bounds[2, 1], q[end])
        s = reshape(s, (:,1))
    end

    return s
end

File: 3_data_fits/gridsamp.py
import numpy as np
from typing import Union, List

def gridsamp(bounds: np.ndarray, q: Union[int, np.ndarray, List[int]]) -> np.ndarray:
    """
    GRIDSAMP  n-dimensional grid over given range

    Parameters
    ----------
    bounds : np.ndarray
        2*n matrix with lower and upper limits
    q : np.ndarray
        n-vector, q(j) is the number of points
        in the j'th direction.
        If q is a scalar, then all q(j) = q

    Returns
    -------
    S : np.ndarray
        m*n array with points, m = prod(q)
    """

    mr, n = np.shape(bounds)
    dr = np.diff(bounds, axis=0)[0]  # difference across rows
    if mr != 2 or any([item < 0 for item in dr]):
        raise Exception('bounds must be an array with two rows and bounds(1,:) <= bounds(2,:)')

    if q.ndim > 1 or any([item <= 0 for item in q]):
        raise Exception('q must be a vector with non-negative elements')

    p = len(q)
    if p == 1:
        q = np.tile(q, (1, n))[0]
    elif p != n:
        raise Exception('length of q must be either 1 or %d' % n)

    # Check for degenerate intervals
    i = np.where(dr == 0)[0]
    if i.size > 0:
        q[i] = 0 * q[i]

    # Recursive computation
    if n > 1:
        a = gridsamp(bounds[:, 1::], q[1::])  # Recursive call
        [m, _] = np.shape(a)
        q = int(q[0])
        s = np.concatenate((np.zeros((m * q, 1)), np.tile(a, (q, 1))), axis=1)
        y = np.linspace(bounds[0, 0], bounds[1, 0], q)

        k = range(m)
        for i in range(q):
            aug = np.tile(y[i], (m, 1))
            aug = np.reshape(aug, s[k, 0].shape)

            s[k, 0] = aug
            k = [item + m for item in k]
    else:
        s = np.linspace(bounds[0, 0], bounds[1, 0], int(q[-1]))
        s = np.transpose([s])

    return s